### ISSUE TITLE
Client with extremely low message rate gets disconnected for unclear reason

### ISSUE THREAD

**Issue by anvouk** [emqx/emqx#11543]
### What happened?

I have an mqtt client which sends 1 messages every hour. After a couple of hours, the client gets kicked out. The broker has message limits but nowhere near the "1 message an hour" limit.

Is this a weird interaction? Am I missing something? what does `reason: [context: {case_clause,{error,{options,{socket_options,{active,100}}}}}` even mean?

Thanks.

### What did you expect to happen?

The client should not get kicked out.

### How can we reproduce it (as minimally and precisely as possible)?

Here's the broker config:

```
mqtt {
    strict_mode = true
    max_inflight = 1
    max_awaiting_rel = 100
    max_mqueue_len = 1000
    mqueue_store_qos0 = false
    use_username_as_clientid = false
    peer_cert_as_username = disabled
    peer_cert_as_clientid = disabled
}
broker {
    session_locking_strategy = quorum
    shared_subscription_strategy = sticky
}
overload_protection {
    enable = true
    backoff_delay = 1
    backoff_gc = false
    backoff_hibernation = true
    backoff_new_conn = true
}
limiter {
    message_in {
        rate = "500/s"
        burst = 10000
    }
    connection {
        rate = "100/s"
        burst = 100
    }
}
persistent_session_store {
    enabled = true
    on_disc = true
    ram_cache = true
    backend {
        type = "builtin"
        messages {
            ram_cache = true
        }
        session {
            ram_cache = true
        }
        session_messages {
            ram_cache = true
        }
    }
    max_retain_undelivered = 24h
}
retainer {
    enable = true
    msg_expiry_interval = 72h
    msg_clear_interval = 1h
    backend {
        type = "built_in_database"
        storage_type = disc
    }
}
listeners {
    ssl {
        default {
            enabled = true
            bind = "0.0.0.0:8883"
            acceptors = 32
            max_connections = 512000
            zone = default
            ssl_options {
                handshake_timeout = 5s
                versions = [tlsv1.3, tlsv1.2]
            }
            limiter.client {
                message_in {
                    initial = 2000
                    capacity = 2000
                    rate = "3/s"
                    low_watermark = 200
                    max_retry_time = 5s
                }
                connection {
                    initial = 10
                    capacity = 10
                    rate = "5/m"
                    low_watermark = 3
                    max_retry_time = 10s
                }
            }
        }
    }
    ws.default.enabled = false
    wss.default.enabled = false
}
```

### Anything else we need to know?

It's a cluster of 3 brokers with only 2 clients connected.

### EMQX version

<details>

```console
$ ./bin/emqx_ctl broker
sysdescr  : EMQX
version   : 5.0.26
datetime  : 2023-08-30T10:22:09.959165909+00:00
uptime    : 22 days, 21 hours, 48 minutes, 43 seconds
```

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
$ uname -a
Linux emqx-0 5.10.179-168.710.amzn2.x86_64 #1 SMP Mon May 22 23:10:22 UTC 2023 x86_64 GNU/Linux
```

</details>

### Log files

Here's the disconnect reason logged from the trace:
<details>
2023-08-30T10:05:28.906215+00:00 [SOCKET] REDACTED@REDACTED:1996 msg: emqx_connection_terminated, reason: [context: {case_clause,{error,{options,{socket_options,{active,100}}}}}, exception: error, stacktrace: [{emqx_connection,process_msg,2,[{file,"emqx_connection.erl"},{line,485}]},{emqx_connection,process_msg,2,[{file,"emqx_connection.erl"},{line,491}]},{emqx_connection,handle_recv,3,[{file,"emqx_connection.erl"},{line,447}]},{proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,236}]}]]
2023-08-30T10:05:28.906386+00:00 [error] crasher: initial call: emqx_connection:init/4, pid: <0.14690.404>, registered_name: [], error: {{case_clause,{error,{options,{socket_options,{active,100}}}}},[{emqx_connection,process_msg,2,[{file,"emqx_connection.erl"},{line,485}]},{emqx_connection,process_msg,2,[{file,"emqx_connection.erl"},{line,491}]},{emqx_connection,handle_recv,3,[{file,"emqx_connection.erl"},{line,447}]},{proc_lib,wake_up,3,[{file,"proc_lib.erl"},{line,236}]}]}, ancestors: [<0.2356.0>,<0.2355.0>,esockd_sup,<0.1989.0>], message_queue_len: 0, messages: [], links: [<0.2356.0>], dictionary: [{guid,{1693387820824324,18148719933794,11}},{send_msg,5},{{publish,<<"REDACTED">>},{allow,1693389928905}},{recv_msg,9},{'$logger_metadata$',#{clientid => <<"REDACTED">>,peername => "REDACTED:1996"}},{authz_keys_q,{[{publish,<<"REDACTED">>}],[]}},{authz_cache_size,1},{recv_pkt,331},{incoming_pubs,9},{'send_msg.qos1',5},{rand_seed,{#{bits => 58,jump => #Fun<rand.3.92093067>,next => #Fun<rand.0.92093067>,type => exsss,uniform => #Fun<rand.1.92093067>,uniform_n => #Fun<rand.2.92093067>},[218432477920937540|228612895014799947]}},{send_pkt,330},{outgoing_bytes,1108},{outgoing_pubs,5},{incoming_bytes,2051},{'recv_msg.qos1',9}], trap_exit: false, status: running, heap_size: 1598, stack_size: 29, reductions: 802651; neighbours:
</details>

**Comment by lafirest** [comment]
@anvouk 
Thanks for your feedback, this is a strange exception that seems like an already fixed bug in Erlang/OTP, https://github.com/erlang/otp/issues/3842.After a preliminary look, only SSL sockets may throw an exception like this `{error,{options,{socket_options,xxx}}}`, but not the `active_n`, we will continue to investigate. 
By the way, we refactored the structure of the configuration for the limiter in 5.1, you could have a try.

**Comment by anvouk** [comment]
The new rate limit config seems quite lacking:
```
listeners.tcp {
    max_conn_rate = "1000/s"
    messages_rate = "1000/s"
    bytes_rate = "1MB/s"
}
```

compared to the old one:
```
listeners.tcp.limiter.client {
    message_in {
        initial = 2000
        capacity = 2000
        rate = "3/s"
        low_watermark = 200
        max_retry_time = 5s
    }
}
```

Have the other fields been removed? I can't find almost anything about these changes...

**Comment by lafirest** [comment]
> The new rate limit config seems quite lacking:
> 
> ```
> listeners.tcp {
>     max_conn_rate = "1000/s"
>     messages_rate = "1000/s"
>     bytes_rate = "1MB/s"
> }
> ```
> 
> compared to the old one:
> 
> ```
> listeners.tcp.limiter.client {
>     message_in {
>         initial = 2000
>         capacity = 2000
>         rate = "3/s"
>         low_watermark = 200
>         max_retry_time = 5s
>     }
> }
> ```
> 
> Have the other fields been removed? I can't find almost anything about these changes...

Yes, the new one aims to simplify the use cost, but it is compatible with the old configuration data.

**Comment by zmstone** [comment]
Hi @anvouk Sorry for missing this out.

The `reason: [context: {case_clause,{error,{options,{socket_options,{active,100}}}}}` is caused by a bug in the code, which is reason why I reoped this issue. Will follow-up with a fix.

However, this bug happens after the socket is already closed. i.e. it's not the reason of your client's disconnection.

There are many reasons for a TCP (or TLS/TCP) connection to close, a common one is after idling for long, the loadbalancers or NAT gateways dropped it.
You might consider sending MQTT PINGREQ from the clients, or enable TCP keepalive from the clients.
